File: sage/tokenizer.py
"""Role-aware tokenizer: Conversation -> (input_ids, attention_mask, pooling_mask)."""

from __future__ import annotations

from dataclasses import dataclass

from sage.conversation import (
    CURRENT_TOKEN,
    ROLE_TOKEN,
    SPECIAL_TOKENS,
    Conversation,
    Role,
)

DEFAULT_BASE_TOKENIZER = "jinaai/jina-embeddings-v2-base-en"
DEFAULT_MAX_LENGTH = 1024


@dataclass
class EncodedConversation:
    input_ids: list[int]
    attention_mask: list[int]
    pooling_mask: list[int]
    length: int

    def validate(self) -> None:
        n = len(self.input_ids)
        assert len(self.attention_mask) == n
        assert len(self.pooling_mask) == n
        assert any(self.pooling_mask)
        for a, p in zip(self.attention_mask, self.pooling_mask, strict=True):
            if p and not a:
                raise AssertionError("pooling_mask covers a pad position")


class SageTokenizer:
    def __init__(
        self,
        base_tokenizer_name: str = DEFAULT_BASE_TOKENIZER,
        *,
        max_length: int = DEFAULT_MAX_LENGTH,
        tokenizer=None,
    ) -> None:
        self.max_length = max_length
        if tokenizer is not None:
            self.tokenizer = tokenizer
        else:
            from transformers import AutoTokenizer

            self.tokenizer = AutoTokenizer.from_pretrained(
                base_tokenizer_name, trust_remote_code=True
            )
        self._added_token_count = self.tokenizer.add_special_tokens(
            {"additional_special_tokens": list(SPECIAL_TOKENS)}
        )

        self.cls_id = self.tokenizer.cls_token_id
        self.sep_id = self.tokenizer.sep_token_id
        self.pad_id = self.tokenizer.pad_token_id
        assert self.cls_id is not None and self.sep_id is not None and self.pad_id is not None

        def _id(tok: str) -> int:
            ids = self.tokenizer.convert_tokens_to_ids([tok])
            assert ids[0] != self.tokenizer.unk_token_id, f"special token {tok} not registered"
            return ids[0]

        self.role_id: dict[Role, int] = {r: _id(ROLE_TOKEN[r]) for r in Role}
        self.current_id: int = _id(CURRENT_TOKEN)

    @property
    def vocab_size(self) -> int:
        return len(self.tokenizer)

    @property
    def added_token_count(self) -> int:
        return self._added_token_count

    def encode(self, conversation: Conversation) -> EncodedConversation:
        ctx_ids: list[list[int]] = [self._encode_text(t.text) for t in conversation.context]
        target_ids = self._encode_text(conversation.target.text)

        budget = self.max_length - 1  # [CLS]
        target_cost = 1 + len(target_ids) + 1
        if target_cost > budget:
            target_ids = target_ids[: budget - 2]
            target_cost = 1 + len(target_ids) + 1

        remaining = budget - target_cost
        has_system = bool(conversation.context) and conversation.context[0].role is Role.SYSTEM
        reserved_system = 0
        if has_system:
            sys_cost = 1 + len(ctx_ids[0]) + 1
            if sys_cost <= remaining:
                reserved_system = sys_cost

        # Keep newest context turns, drop oldest first.
        start = 1 if has_system else 0
        kept: list[int] = []
        cost = 0
        for i in range(len(conversation.context) - 1, start - 1, -1):
            turn_cost = 1 + len(ctx_ids[i]) + 1
            if cost + turn_cost + reserved_system <= remaining:
                kept.append(i)
                cost += turn_cost
            else:
                break
        kept.reverse()
        if has_system and reserved_system > 0:
            kept = [0, *kept]

        input_ids: list[int] = [self.cls_id]
        pooling_mask: list[int] = [0]
        for i in kept:
            turn = conversation.context[i]
            input_ids.append(self.role_id[turn.role])
            pooling_mask.append(0)
            input_ids.extend(ctx_ids[i])
            pooling_mask.extend([0] * len(ctx_ids[i]))
            input_ids.append(self.sep_id)
            pooling_mask.append(0)

        input_ids.append(self.current_id)
        pooling_mask.append(1)
        input_ids.extend(target_ids)
        pooling_mask.extend([1] * len(target_ids))
        input_ids.append(self.sep_id)
        pooling_mask.append(0)

        attention_mask = [1] * len(input_ids)
        pad = self.max_length - len(input_ids)
        if pad > 0:
            input_ids.extend([self.pad_id] * pad)
            attention_mask.extend([0] * pad)
            pooling_mask.extend([0] * pad)

        enc = EncodedConversation(input_ids, attention_mask, pooling_mask, sum(attention_mask))
        enc.validate()
        return enc

    def encode_batch(self, conversations: list[Conversation]) -> dict[str, list[list[int]]]:
        encoded = [self.encode(c) for c in conversations]
        return {
            "input_ids": [e.input_ids for e in encoded],
            "attention_mask": [e.attention_mask for e in encoded],
            "pooling_mask": [e.pooling_mask for e in encoded],
        }

    def _encode_text(self, text: str) -> list[int]:
        return self.tokenizer.encode(text, add_special_tokens=False)

File: rust/sage-rs/src/tokenizer.rs
//! Role-aware tokenizer mirroring `sage/tokenizer.py`.
//!
//! Takes a HuggingFace `tokenizers::Tokenizer` and produces the three parallel
//! integer sequences SAGE needs: `input_ids`, `attention_mask`, `pooling_mask`.
//!
//! Truncation policy (must stay aligned with Python):
//! 1. Drop oldest non-system context turns first.
//! 2. Pin a system turn at position 0 unless keeping it forces target truncation.
//! 3. Right-truncate the target if it alone exceeds the budget.
//! 4. Always keep [CLS] and the [CURRENT] marker.

use tokenizers::{AddedToken, Tokenizer};

use crate::conversation::{Conversation, Role, CURRENT_TOKEN, SPECIAL_TOKENS};
use crate::error::{Result, SageError};

pub(crate) struct Encoded {
    pub input_ids: Vec<i64>,
    pub attention_mask: Vec<i64>,
    pub pooling_mask: Vec<i64>,
}

pub(crate) struct SageTokenizer {
    inner: Tokenizer,
    max_length: usize,
    cls_id: u32,
    sep_id: u32,
    pad_id: u32,
    user_id: u32,
    char_id: u32,
    system_id: u32,
    current_id: u32,
}

impl SageTokenizer {
    pub fn new(mut inner: Tokenizer, max_length: usize) -> Result<Self> {
        // Register special tokens. Adding an already-present special token is a no-op.
        let added: Vec<AddedToken> = SPECIAL_TOKENS
            .iter()
            .map(|t| AddedToken::from(t.to_string(), true))
            .collect();
        inner.add_special_tokens(&added);

        let lookup = |name: &'static str| -> Result<u32> {
            inner
                .token_to_id(name)
                .ok_or(SageError::MissingSpecialToken(name))
        };

        Ok(Self {
            cls_id: lookup("[CLS]")?,
            sep_id: lookup("[SEP]")?,
            pad_id: lookup("[PAD]")?,
            user_id: lookup("[USER]")?,
            char_id: lookup("[CHAR]")?,
            system_id: lookup("[SYSTEM]")?,
            current_id: lookup(CURRENT_TOKEN)?,
            inner,
            max_length,
        })
    }

    pub fn max_length(&self) -> usize {
        self.max_length
    }

    fn role_id(&self, role: Role) -> u32 {
        match role {
            Role::User => self.user_id,
            Role::Char => self.char_id,
            Role::System => self.system_id,
        }
    }

    fn encode_text(&self, text: &str) -> Result<Vec<u32>> {
        let encoding = self.inner.encode(text, false)?;
        Ok(encoding.get_ids().to_vec())
    }

    pub fn encode(&self, conv: &Conversation) -> Result<Encoded> {
        // 1. Tokenize every turn's text in one pass.
        let ctx = conv.context();
        let mut ctx_ids: Vec<Vec<u32>> = Vec::with_capacity(ctx.len());
        for t in ctx {
            ctx_ids.push(self.encode_text(&t.text)?);
        }
        let mut target_ids = self.encode_text(&conv.target().text)?;

        // 2. Budget: max_length - 1 for [CLS].
        let budget = self.max_length.saturating_sub(1);
        let mut target_cost = 1 + target_ids.len() + 1; // [CURRENT] ... [SEP]
        if target_cost > budget {
            let max_target_text = budget.saturating_sub(2);
            target_ids.truncate(max_target_text);
            target_cost = 1 + target_ids.len() + 1;
        }
        let remaining = budget.saturating_sub(target_cost);

        // Is the first context turn a SYSTEM? (pinned)
        let has_pinned_system = matches!(ctx.first().map(|t| t.role), Some(Role::System));

        // Reserve space for a pinned system turn if it fits.
        let mut reserved_system: usize = 0;
        if has_pinned_system {
            let sys_cost = 1 + ctx_ids[0].len() + 1;
            if sys_cost <= remaining {
                reserved_system = sys_cost;
            }
        }

        // Walk non-system context turns newest → oldest, greedy keep.
        let non_system_start: usize = if has_pinned_system { 1 } else { 0 };
        let mut kept_indices: Vec<usize> = Vec::new();
        let mut cost_of_kept: usize = 0;

        if ctx.len() > non_system_start {
            for i in (non_system_start..ctx.len()).rev() {
                let turn_cost = 1 + ctx_ids[i].len() + 1;
                if cost_of_kept + turn_cost + reserved_system <= remaining {
                    kept_indices.push(i);
                    cost_of_kept += turn_cost;
                } else {
                    break;
                }
            }
            kept_indices.reverse();
        }
        if has_pinned_system && reserved_system > 0 {
            let mut with_sys = Vec::with_capacity(kept_indices.len() + 1);
            with_sys.push(0usize);
            with_sys.extend(kept_indices);
            kept_indices = with_sys;
        }

        // 3. Assemble.
        let mut input_ids: Vec<i64> = Vec::with_capacity(self.max_length);
        let mut pooling_mask: Vec<i64> = Vec::with_capacity(self.max_length);

        input_ids.push(self.cls_id as i64);
        pooling_mask.push(0);

        for i in &kept_indices {
            let role = ctx[*i].role;
            input_ids.push(self.role_id(role) as i64);
            pooling_mask.push(0);
            for tok in &ctx_ids[*i] {
                input_ids.push(*tok as i64);
                pooling_mask.push(0);
            }
            input_ids.push(self.sep_id as i64);
            pooling_mask.push(0);
        }

        // Target: [CURRENT] ... text ... [SEP]
        input_ids.push(self.current_id as i64);
        pooling_mask.push(1); // [CURRENT] is included in the pool
        for tok in &target_ids {
            input_ids.push(*tok as i64);
            pooling_mask.push(1);
        }
        input_ids.push(self.sep_id as i64);
        pooling_mask.push(0);

        // attention_mask starts as all-1
        let mut attention_mask: Vec<i64> = vec![1; input_ids.len()];

        // 4. Pad to max_length.
        if input_ids.len() < self.max_length {
            let pad = self.max_length - input_ids.len();
            input_ids.extend(std::iter::repeat(self.pad_id as i64).take(pad));
            attention_mask.extend(std::iter::repeat(0).take(pad));
            pooling_mask.extend(std::iter::repeat(0).take(pad));
        }

        debug_assert_eq!(input_ids.len(), self.max_length);
        debug_assert_eq!(attention_mask.len(), self.max_length);
        debug_assert_eq!(pooling_mask.len(), self.max_length);

        Ok(Encoded { input_ids, attention_mask, pooling_mask })
    }
}
